File: sample/sample-java/src/test/java/com/quantipixels/ogiri/samples/java/jdbc/JdbcSampleTokenRepositoryTest.java
/*
 * Copyright (c) 2025 Quanti Pixels
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */
package com.quantipixels.ogiri.samples.java.jdbc;

import static org.junit.jupiter.api.Assertions.*;

import com.quantipixels.ogiri.samples.java.Application;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;

@SpringBootTest(classes = Application.class, webEnvironment = SpringBootTest.WebEnvironment.NONE)
@ActiveProfiles("jdbc")
class JdbcSampleTokenRepositoryTest {

  @Autowired private JdbcSampleTokenRepository tokenRepository;

  private static final long TEST_USER_ID = 100L;
  private static final String TEST_CLIENT = "test-client";
  private static final String TEST_TOKEN = "hashed-token-123";

  @BeforeEach
  void setUp() {
    tokenRepository.deleteByUserId(TEST_USER_ID);
    tokenRepository.deleteByUserId(101L);
    tokenRepository.deleteByUserId(102L);
  }

  @Test
  void shouldSaveAndRetrieveTokenByUserAndClient() {
    JdbcSampleToken token = createToken(TEST_USER_ID, TEST_CLIENT, TEST_TOKEN);
    tokenRepository.save(token);

    Optional<JdbcSampleToken> retrieved =
        tokenRepository.findByUserIdAndClient(TEST_USER_ID, TEST_CLIENT);

    assertTrue(retrieved.isPresent());
    assertEquals(TEST_USER_ID, retrieved.get().getUserId());
    assertEquals(TEST_CLIENT, retrieved.get().getClient());
    assertEquals(TEST_TOKEN, retrieved.get().getToken());
  }

  @Test
  void shouldReturnEmptyOptionalForNonExistentToken() {
    Optional<JdbcSampleToken> retrieved = tokenRepository.findByUserIdAndClient(999L, "none");
    assertFalse(retrieved.isPresent());
  }

  @Test
  void shouldFindAllTokensForUserAndExcludeOtherUsers() {
    tokenRepository.save(createToken(TEST_USER_ID, "client-1", "token-1"));
    tokenRepository.save(createToken(TEST_USER_ID, "client-2", "token-2"));
    tokenRepository.save(createToken(101L, "client-3", "token-3"));

    List<JdbcSampleToken> tokens = tokenRepository.findByUserIdOrderByUpdatedAtDesc(TEST_USER_ID);

    assertEquals(2, tokens.size());
    List<String> clients = tokens.stream().map(JdbcSampleToken::getClient).toList();
    assertTrue(clients.contains("client-1"));
    assertTrue(clients.contains("client-2"));
  }

  @Test
  void shouldFindExpiredTokens() {
    Instant now = Instant.now();

    JdbcSampleToken expired = createToken(TEST_USER_ID, "expired-client", TEST_TOKEN);
    expired.setExpiryAt(now.minus(1, ChronoUnit.HOURS));

    JdbcSampleToken valid = createToken(TEST_USER_ID, "valid-client", TEST_TOKEN);
    valid.setExpiryAt(now.plus(1, ChronoUnit.HOURS));

    tokenRepository.save(expired);
    tokenRepository.save(valid);

    List<JdbcSampleToken> expiredTokens = tokenRepository.findByExpiryAtBefore(now);

    assertTrue(expiredTokens.stream().anyMatch(t -> t.getClient().equals("expired-client")));
    assertTrue(expiredTokens.stream().noneMatch(t -> t.getExpiryAt().isAfter(now)));
  }

  @Test
  void shouldDeleteTokenByUserAndClient() {
    tokenRepository.save(createToken(TEST_USER_ID, TEST_CLIENT, TEST_TOKEN));

    tokenRepository.deleteByUserIdAndClient(TEST_USER_ID, TEST_CLIENT);

    assertFalse(tokenRepository.findByUserIdAndClient(TEST_USER_ID, TEST_CLIENT).isPresent());
  }

  @Test
  void shouldDeleteSelectedTokensByClientList() {
    tokenRepository.save(createToken(102L, "client-1", "token-1"));
    tokenRepository.save(createToken(102L, "client-2", "token-2"));
    tokenRepository.save(createToken(102L, "client-3", "token-3"));

    tokenRepository.deleteByUserIdAndClientIn(102L, List.of("client-1", "client-2"));

    List<JdbcSampleToken> remaining = tokenRepository.findByUserIdOrderByUpdatedAtDesc(102L);
    assertEquals(1, remaining.size());
    assertEquals("client-3", remaining.get(0).getClient());
  }

  private JdbcSampleToken createToken(long userId, String client, String tokenHash) {
    JdbcSampleToken token = new JdbcSampleToken();
    token.setUserId(userId);
    token.setClient(client);
    token.setToken(tokenHash);
    token.setExpiryAt(Instant.now().plus(1, ChronoUnit.HOURS));
    return token;
  }
}

File: sample/sample-kotlin/src/test/kotlin/com/quantipixels/ogiri/samples/kotlin/jdbc/JdbcSampleTokenRepositoryTest.kt
/*
 * Copyright (c) 2025 Quanti Pixels
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */
package com.quantipixels.ogiri.samples.kotlin.jdbc

import com.quantipixels.ogiri.samples.kotlin.Application
import java.time.Instant
import java.time.temporal.ChronoUnit
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles

@SpringBootTest(classes = [Application::class], webEnvironment = SpringBootTest.WebEnvironment.NONE)
@ActiveProfiles("jdbc")
class JdbcSampleTokenRepositoryTest {

  @Autowired private lateinit var tokenRepository: JdbcSampleTokenRepository

  private val testUserId = 100L
  private val testClient = "test-client"
  private val testToken = "hashed-token-123"

  @BeforeEach
  fun setUp() {
    tokenRepository.deleteByUserId(testUserId)
    tokenRepository.deleteByUserId(101L)
    tokenRepository.deleteByUserId(102L)
  }

  @Test
  fun `should save and retrieve token by user and client`() {
    val token = createToken(testUserId, testClient, testToken)
    tokenRepository.save(token)

    val retrieved = tokenRepository.findByUserIdAndClient(testUserId, testClient)

    assertTrue(retrieved.isPresent)
    assertEquals(testUserId, retrieved.get().userId)
    assertEquals(testClient, retrieved.get().client)
    assertEquals(testToken, retrieved.get().token)
  }

  @Test
  fun `should return empty Optional for non-existent token`() {
    val retrieved = tokenRepository.findByUserIdAndClient(999L, "non-existent")
    assertFalse(retrieved.isPresent)
  }

  @Test
  fun `should find all tokens for user and exclude tokens from other users`() {
    val token1 = createToken(testUserId, "client-1", "token-1")
    val token2 = createToken(testUserId, "client-2", "token-2")
    val otherUserToken = createToken(101L, "client-3", "token-3")

    tokenRepository.save(token1)
    tokenRepository.save(token2)
    tokenRepository.save(otherUserToken)

    val tokens = tokenRepository.findByUserIdOrderByUpdatedAtDesc(testUserId)

    assertEquals(2, tokens.size)
    val clients = tokens.map { it.client }
    assertTrue(clients.contains("client-1"))
    assertTrue(clients.contains("client-2"))
  }

  @Test
  fun `should find expired tokens`() {
    val now = Instant.now()

    val expiredToken =
        createToken(testUserId, "expired-client", testToken).apply {
          expiryAt = now.minus(1, ChronoUnit.HOURS)
        }
    val validToken =
        createToken(testUserId, "valid-client", testToken).apply {
          expiryAt = now.plus(1, ChronoUnit.HOURS)
        }

    tokenRepository.save(expiredToken)
    tokenRepository.save(validToken)

    val expired = tokenRepository.findByExpiryAtBefore(now)

    assertTrue(expired.any { it.client == "expired-client" })
    assertTrue(expired.none { it.expiryAt.isAfter(now) })
  }

  @Test
  fun `should delete token by user and client`() {
    tokenRepository.save(createToken(testUserId, testClient, testToken))

    tokenRepository.deleteByUserIdAndClient(testUserId, testClient)

    assertFalse(tokenRepository.findByUserIdAndClient(testUserId, testClient).isPresent)
  }

  @Test
  fun `should delete selected tokens by client list`() {
    tokenRepository.save(createToken(102L, "client-1", "token-1"))
    tokenRepository.save(createToken(102L, "client-2", "token-2"))
    tokenRepository.save(createToken(102L, "client-3", "token-3"))

    tokenRepository.deleteByUserIdAndClientIn(102L, listOf("client-1", "client-2"))

    val remaining = tokenRepository.findByUserIdOrderByUpdatedAtDesc(102L)
    assertEquals(1, remaining.size)
    assertEquals("client-3", remaining[0].client)
  }

  private fun createToken(userId: Long, client: String, token: String): JdbcSampleToken =
      JdbcSampleToken().apply {
        this.userId = userId
        this.client = client
        this.token = token
        this.expiryAt = Instant.now().plus(1, ChronoUnit.HOURS)
      }
}
